group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1040,6 +1051,7 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
             .maybeUpdateRebalanceTimeoutMs(ofSentinel(rebalanceTimeoutMs))
             .maybeUpdateServerAssignorName(Optional.ofNullable(assignorName))
             .maybeUpdateSubscribedTopicNames(Optional.ofNullable(subscribedTopicNames))
+            .maybeUpdateSubscribedTopicRegex(Optional.ofNullable(subscribedTopicRegex))
```

Comment:
I think that we miss a few fundamental things in `consumerGroupHeartbeat`. Let me explain. The `consumerGroupHeartbeat` is structured in 3 parts.
1) We update the member and its subscriptions.
2) We compute the new target assignment if needed.
3) We reconcile the member.

I think that in 1), we need to update the subscription for the member as you do here. However, we also need to verify it before storing it and we also need to update the subscription metadata if the regex was changed. See L1080. In step 2), we also need to change the logic to include the topic matching the regex. See L1115. Step 3) is fine as it is.

We also need a mechanism to periodically refresh the regexes in order to catch new topics or deleted topics. What was your plan for this? I thought that we could piggy back the mechanism to refresh the subscription metadata (L1076) to also refresh the regexes.

I think that we also need to store the resolved regular expressions somehow. I mean a mapping from the regex (as string) to the matching topics because we need this for step 2). For this, I was considering whether we could just use a LRU cache. What do you think?

Reply:
@dajac, +1 on using LRU cache, it would also save us from unnecessary assignment computations when member rejoining